Fix endless recovery problem

TaoYang526 opened the pull request:
After restarting scheduler with a running application which can be seen by the scheduler,  I found the recovery process is endless and always wait for the application whose state is Running. 
According to the code in Context#waitForAppRecovery, the condition to finish this recovery is that app state should be Accepted, but app state may also have a chance to be transitioned to Running and other states(like Completed, Killed, etc.) here.
Perhaps we can expect that recovery process of this app is finished when app state is not New and not Recovering?
@yangwwei @wilfred-s  please help to review this PR, thanks.

yangwwei commented:
Hi @TaoYang526 

According to the design doc: https://github.com/cloudera/yunikorn-core/blob/master/docs/design/resilience.md#applications.  For applications, shim submits existing apps using AddApplication API call, and once the app is accepted by scheduler-core, its state will be "ACCEPTED". And this will be the state app is holding on because the scheduling is not going to be triggered before the recovery process ends. I am wondering how the app can reach RUNNING state during the recovery process. Do you have any more info about that?

TaoYang526 commented:
@yangwwei 
I have tried to go through the recovery flow of app state from New to Running as follows, hope it can help:
(1) trigger recovery: KubernetesShim#recoverSchedulerState -> Context#WaitForRecovery -> Context#waitForAppRecovery -> Context#getOrCreateApplication -> cache.NewApplication -> app.state==New
(2) transition to Recovering: Context#waitForAppRecovery -> dispatcher.Dispatch(NewSimpleApplicationEvent(appId, events.RecoverApplication)) -- app handle events.RecoverApplication --> app state machine:  transition to Recovering
(3) transition to Accepted: Application#handleRecoverApplicationEvent -> RMProxy#Update  --> CacheEventHandler#HandleEvent(&cacheevent.RMUpdateRequestEvent{...}) --> ClusterInfo#handleRMEvents -> ClusterInfo#processRMUpdateEvent -> ClusterInfo#processApplicationUpdateFromRMUpdate -> SchedulerEventHandler#HandleEvent(&schedulerevent.SchedulerApplicationsUpdateEvent{...}) -> Scheduler#handleSchedulerEvent -> Scheduler#processApplicationUpdateEvent --> RMProxyEventHandler.HandleEvent(&rmevent.RMApplicationUpdateEvent{...}) -> RMProxy#processApplicationUpdateEvent -> RMProxy#processUpdateResponse -> AsyncRMCallback#RecvUpdateResponse -> dispatcher.Dispatch(cache.NewSimpleApplicationEvent(appId, events.AcceptApplication)) --> app state machine: transition to Accepted
(4) transition to Running: Application#handleAcceptApplicationEvent -> dispatcher.Dispatch(NewRunApplicationEvent(appId)) -> Application#handle -> app.sm.Event(...) ->  app state machine: transition to Running

Key logs:

```
2019-12-10T07:59:12.994Z    INFO    shim/scheduler.go:152   recovering scheduler states
2019-12-10T07:59:13.994Z    INFO    cache/context_recovery.go:74    start to recover the app    {"appId": "MyOwnApplicationId-4"}
2019-12-10T07:59:13.994Z    INFO    cache/context_recovery.go:74    start to recover the app    {"appId": "MyOwnApplicationId-4"}
2019-12-10T07:59:13.994Z    INFO    cache/context_recovery.go:74    start to recover the app    {"appId": "MyOwnApplicationId-4"}
2019-12-10T07:59:13.994Z    INFO    cache/context_recovery.go:74    start to recover the app    {"appId": "MyOwnApplicationId-4"}
2019-12-10T07:59:13.994Z    INFO    cache/context_recovery.go:74    start to recover the app    {"appId": "MyOwnApplicationId-4"}
2019-12-10T07:59:13.994Z    INFO    cache/context_recovery.go:74    start to recover the app    {"appId": "MyOwnApplicationId-4"}
2019-12-10T07:59:13.994Z    INFO    cache/context_recovery.go:74    start to recover the app    {"appId": "MyOwnApplicationId-4"}
2019-12-10T07:59:13.994Z    INFO    cache/context_recovery.go:74    start to recover the app    {"appId": "MyOwnApplicationId-4"}
2019-12-10T07:59:13.994Z    INFO    cache/context_recovery.go:74    start to recover the app    {"appId": "MyOwnApplicationId-4"}
2019-12-10T07:59:13.994Z    INFO    cache/context_recovery.go:74    start to recover the app    {"appId": "MyOwnApplicationId-4"}
2019-12-10T07:59:13.994Z    INFO    cache/context_recovery.go:86    appInfo {"appId": "MyOwnApplicationId-4", "state": "New"}
2019-12-10T07:59:13.994Z    DEBUG   cache/application.go:126    application state transition    {"appId": "MyOwnApplicationId-4", "preState": "New", "pendingEvent": "RecoverApplication"}
2019-12-10T07:59:13.994Z    INFO    cache/application.go:287    handle app recovering   {"app": "applicationId: MyOwnApplicationId-4, queue: root.sandbox, partition: default, totalNumOfTasks: 0, currentState: Recovering", "clusterId": "mycluster"}
2019-12-10T07:59:13.994Z    DEBUG   cache/application.go:135    application state transition    {"appId": "MyOwnApplicationId-4", "postState": "Recovering"}
2019-12-10T07:59:13.994Z    INFO    cache/partition_info.go:346 adding app to partition {"appId": "MyOwnApplicationId-4", "queue": "root.sandbox", "partitionName": "[mycluster]default"}
2019-12-10T07:59:13.995Z    INFO    cache/partition_info.go:370 app added to partition  {"appId": "MyOwnApplicationId-4", "partitionName": "[mycluster]default"}
2019-12-10T07:59:13.995Z    DEBUG   scheduler/scheduler.go:190  enqueued event  {"eventType": "*schedulerevent.SchedulerApplicationsUpdateEvent", "event": {"AddedApplications":[{"ApplicationId":"MyOwnApplicationId-4","Partition":"[mycluster]default","QueueName":"root.sandbox","SubmissionTime":1575964753994957911}],"RemovedApplications":null}, "currentQueueSize": 0}
2019-12-10T07:59:13.995Z    DEBUG   cache/application_state.go:89   app state transition    {"app": "MyOwnApplicationId-4", "source": "New", "destination": "Accepted", "event": "AcceptApplication"}
2019-12-10T07:59:13.995Z    DEBUG   rmproxy/rmproxy.go:62   enqueue event   {"event": {"RMId":"mycluster","AcceptedApplications":[{"applicationId":"MyOwnApplicationId-4"}],"RejectedApplications":[]}, "currentQueueSize": 0}
2019-12-10T07:59:13.994Z    DEBUG   cache/cluster_info.go:131   enqueued event  {"eventType": "*cacheevent.RMUpdateRequestEvent", "event": {"Request":{"rmId":"mycluster","newApplications":[{"applicationId":"MyOwnApplicationId-4","queueName":"root.sandbox","partitionName":"[mycluster]default","ugi":{"user":"default"},"tags":{"namespace":"blink"}}]}}, "currentQueueSize": 0}
2019-12-10T07:59:13.995Z    INFO    callback/scheduler_callback.go:40   callback received   {"updateResponse": "acceptedApplications:<applicationId:\"MyOwnApplicationId-4\" > "}
2019-12-10T07:59:13.995Z    INFO    callback/scheduler_callback.go:67   callback: response to accepted application  {"appId": "MyOwnApplicationId-4"}
2019-12-10T07:59:13.995Z    DEBUG   cache/application.go:126    application state transition    {"appId": "MyOwnApplicationId-4", "preState": "Recovering", "pendingEvent": "AcceptApplication"}
2019-12-10T07:59:13.995Z    DEBUG   cache/application.go:135    application state transition    {"appId": "MyOwnApplicationId-4", "postState": "Accepted"}
2019-12-10T07:59:13.995Z    DEBUG   cache/application.go:126    application state transition    {"appId": "MyOwnApplicationId-4", "preState": "Accepted", "pendingEvent": "RunApplication"}
2019-12-10T07:59:13.995Z    DEBUG   cache/application.go:135    application state transition    {"appId": "MyOwnApplicationId-4", "postState": "Running"}
2019-12-10T07:59:14.994Z    INFO    cache/context_recovery.go:86    appInfo {"appId": "MyOwnApplicationId-4", "state": "Running"}
2019-12-10T07:59:15.995Z    INFO    cache/context_recovery.go:86    appInfo {"appId": "MyOwnApplicationId-4", "state": "Running"}
```

yangwwei commented:
Oh, this is a regression after #57 
Once app is accepted, it now automatically transits to Running. The old behavior will be staying in Accepted until the scheduling process kicks in. Now I see why it is failing.

TaoYang526 commented:
@yangwwei  Yes, it does. Is it better to remove callback logic for events.AcceptApplication? So that this state transition ( from 'Accepted' to 'Running') can only depends on the scheduling process.

yangwwei commented:
Hi @TaoYang526 
I can't recall exact reason why I have added this for #57 , it looks like a dup of the one in scheduling process. Maybe we can remove it. But please make sure all unit tests and smoke tests are working.

TaoYang526 commented:
@yangwwei  I have removed the callback logic of AcceptApplication event which is useless as we thought, and found the failure of TestApplicationSchedulingWithoutDuplicateEvents
 since it counts the RunApplication events for validation via dispatcher which is only implemented in the callback logic of AcceptApplication event,  I removed this UT as well since it's no longer applicable and no need for current logic.
Moreover, I found TestAppRecovery is not correct as it calls Context#waitForAppRecovery twice, it doesn't work as expect in the second call (app state has changed to Recovering in first call so there is no to-recovery apps). Now this UT has been corrected and can validate this recovery problem fine.

wilfred-s commented on pkg/cache/context_recovery_test.go:
```diff
@@ -269,15 +269,23 @@ func TestAppRecovery(t *testing.T) {
+			}
+			return false
+		}, 100*time.Millisecond, 3*time.Second); err != nil {
+			t.Fatal("unexpected app states")
```
Can we add what the states are etc so we know why we failed

TaoYang526 replied:
Sure.

wilfred-s commented on pkg/cache/context_recovery_test.go:
```diff
@@ -269,15 +269,23 @@ func TestAppRecovery(t *testing.T) {
 	podLister.AddPod(&pod2)
 	podLister.AddPod(&pod3)
```
We have three pods, and three apps: if the pod/app does not impact testing we must remove it. If it is crucial for the test we need to check that it is created (or not) etc.

TaoYang526 replied:
Make sense, I'll add validations for apps at last, pods in recovery process have not been handled yet so there's no check for them.

wilfred-s replied:
Nothing happens with pod 3, we do not check that nor can we so we should not have it in the code.
For pod1 and pod2: if one fails both fail. I would be highly surprised if that did not happen. Having 2 in the code thus does not add anything to the test.

I would really simplify the test to just have one app and make sure it works.
If you want to test the non recovery (i.e. the app3 case) create a new test for that and check that the app never gets to a recovering state.

TaoYang526 replied:
> I would really simplify the test to just have one app and make sure it works.

Make sense to me, I will update it later.

> If you want to test the non recovery (i.e. the app3 case) create a new test for that and check that the app never gets to a recovering state.

How about to call `context.GetApplication("app2")` and make sure that it will return error("application app3 is not found in context") at last of this test? I think this can prove that nothing happens with pod3.

TaoYang526 commented:
Thanks @wilfred-s for the review. 

> why are we testing recovery of two exactly the same apps? The test does not check anything special and the pod1 and pod2 are exactly the same (nodes are ignored)

I guess the purpose of this UT is to test if 2 apps already scheduled before (in total 3 apps) can be recovered as expect, so that we can ignore the details of apps/pods which are not cared for this UT.

> Can we have negative testing (failure cases) also or is that impossible because failures only occur on timeout of processing?

Another possible failures are the app states after recovery, checks for them have been added in the latest commit.

yangwwei commented:
well, `context_recovery_test.go` is very specific to verify functions in `context_recovery.go`, not like any kind of e2e test. That's why we mocked `podLister`. For more tests, we have some in `scheduler_recovery_test.go`, that stimulates the whole recovery context.

The change looks good to me now, @wilfred-s , once your comments are addressed, please help to merge this PR. Thanks!

yangwwei commented:
Hi @TaoYang526 can you pls make changes and update the PR today? It will be good if we can get this merged today. Thanks!

TaoYang526 commented:
hi, @wilfred-s I have updated this PR, could you pls help to review it again?
